### ISSUE TITLE
Sensitive outputs shouldn't be exposed when using terraform.OutputRequired

### ISSUE THREAD

**Issue by soerenmartius** [gruntwork-io/terratest#476]
Hi,

We are working on a test-case that creates some IAM users and access keys that we use for authenticating with some AWS resources within that test. To be able to use the credentials we pull them from the outputs,  e.G:

**outputs.tf:**
```
output "aws_iam_access_key_secret" {
  description = "The acccess key secret."
  value       = aws_iam_access_key.docker.secret
  sensitive   = true
}
```

**test.go:**
```
accessKeySecret := terraform.OutputRequired(t, terraformOptions, 'aws_iam_access_key_secret')
```

Using the helper methods to retrieve outputs with terraform will dump the outputs to the logs. I believe we should align the behavior of terratest with terraform and only write `<sensitive>` instead of the actual values to the logs.

**Comment by brikis98** [comment]
Oh, good point! Would you be up for a PR to fix this?

**Comment by soerenmartius** [comment]
Ok, so running `terraform output` will mask any outputs that have `sensitive = true`. If one calls `terraform output` on a specific output ( e.g. `terraform output aws_iam_access_key_secret` terraform returns the value of the output as clear text, which also makes kind of sense to me also.

It seems to me that there is no easy way to be aware if we deal with sensitive data or when calling `terraform output` on a specific output. The log line is written [command.go](https://github.com/gruntwork-io/terratest/blob/master/modules/shell/command.go#L156) file. It might be a reasonable and simpler approach to make the logging of the output configurable.

wdyt @brikis98 ?

**Comment by brikis98** [comment]
Try `terraform output -json`. I believe that shows you which ones are sensitive.

**Comment by soerenmartius** [comment]
yeah sure, `terraform output` also shows you which ones are sensitive. That isn't very helpful though since the outputs methods in terratest would end up calling `terraform output key`, which will always return clear text.

**Comment by brikis98** [comment]
Yea, what I meant was that the `terraform.Output(xxx)` helper would always run `terraform output -json` with no logging, read output `xxx` from the JSON, and only log what it saw if it wasn't sensitive. So it's making the logging optional in `command.go` as you said, but also making the `terraform.Output` helper in how it uses it.

**Comment by soerenmartius** [comment]
Thank you, that makes sense to me! I will take a proper look in a bit.

**Comment by burythehammer** [comment]
We've been seeing this issue as well. We apply some terraform, and then use terratest's output functions to fetch out a randomly generated password. We need the password to ensure that the applied resources have been set up, but by getting the output, it also prints out the password to our logs - not ideal.

It feels like by default logging should be turned off for any sensitive output.

Has there been any progress on this issue?

**Comment by yorinasub17** [comment]
Here is a workaround for this that we recently identified:

```go
func fetchSensitiveOutput(t *testing.T, options *terraform.Options, name string) string {
    defer func() {
        options.Logger = nil
    }()
    options.Logger = logger.Discard  // import path is github.com/gruntwork-io/terratest/modules/logger
    return terraform.Output(t, options, name)
}
```

**Comment by estahn** [comment]
I started using terratest to test my Kubernetes Webhook ([k8s-image-swapper](https://github.com/estahn/k8s-image-swapper)). The IAM keys and docker credentials were exposed in the logs, which is not ideal for CI purposes. I came up with the following solution:

**Usage**:
```golang
func TestHelmDeployment(t *testing.T) {
	workingDir, _ := filepath.Abs("..")

	awsAccountID := os.Getenv("AWS_ACCOUNT_ID")
	awsRegion := os.Getenv("AWS_DEFAULT_REGION")
	awsAccessKeyID := os.Getenv("AWS_ACCESS_KEY_ID")
	awsSecretAccessKey := os.Getenv("AWS_SECRET_ACCESS_KEY")
	ecrRegistry := awsAccountID + ".dkr.ecr." + awsRegion + ".amazonaws.com"
	ecrRepository := "docker.io/library/nginx"

	logger.Default = logger.New(newSensitiveLogger(
		logger.Default,
		[]*regexp.Regexp{
			regexp.MustCompile(awsAccountID),
			regexp.MustCompile(awsAccessKeyID),
			regexp.MustCompile(awsSecretAccessKey),
			regexp.MustCompile(`(--docker-password=)\S+`),
		},
	))

    ...
```

**Implementation:**:

```golang

type sensitiveLogger struct{
	logger logger.TestLogger
	patterns []*regexp.Regexp
}

func newSensitiveLogger(logger *logger.Logger, patterns []*regexp.Regexp) *sensitiveLogger {
	return &sensitiveLogger{
		logger: logger,
		patterns: patterns,
	}
}

func (l *sensitiveLogger) Logf(t terratesttesting.TestingT, format string, args ...interface{}) {
	var redactedArgs []interface{}

	obfuscateWith := "$1*******"

	redactedArgs = args

	for _, pattern := range l.patterns {
		for i, arg := range redactedArgs {
			switch arg := arg.(type) {
			case string:
				redactedArgs[i] = pattern.ReplaceAllString(arg, obfuscateWith)
			case []string:
				var result []string
				for _, s := range arg {
					result = append(result, pattern.ReplaceAllString(s, obfuscateWith))
				}
				redactedArgs[i] = result
			default:
				panic("type needs implementation")
			}
		}
	}

	l.logger.Logf(t, format, redactedArgs...)
}
```

@brikis98 What are your thoughts on this? Do you think this is something we can add to terratest itself?

**Comment by james03160927** [comment]
Seems like a PR was merged for this issue. https://github.com/microsoft/retina/pull/1286
